### PR TITLE
Pipe: support replace and modify mode for alter pipe sql

### PR DESCRIPTION
**Pull request by VGalaxies**
## Description

As title.

<hr>

This PR has:
- [x] been self-reviewed.
    - [ ] concurrent read
    - [ ] concurrent write
    - [ ] concurrent read and write 
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. 
- [ ] added or updated version, __license__, or notice information
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [x] added integration tests.
- [x] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: integration-test/src/test/java/org/apache/iotdb/pipe/it/IoTDBPipeAlterIT.java [hunk 1]
```diff
@@ -103,16 +104,62 @@ public void testBasicPipeAlter() throws Exception {
...
+          showPipeResult
+              .get(0)
+              .pipeConnector
+              .contains(String.format("node-urls=%s", receiverDataNode.getIpAndPortString())));
+      // alter pipe will keep running status
+      Assert.assertEquals("RUNNING", showPipeResult.get(0).state);
+      // alter pipe will reset pipe creation time
+      Assert.assertTrue(showPipeResult.get(0).creationTime > lastCreationTime);
       // alter pipe will clear exception messages
       Assert.assertEquals("", showPipeResult.get(0).exceptionMessage);
```
Comment (Caideyipi) [inline integration-test/src/test/java/org/apache/iotdb/pipe/it/IoTDBPipeAlterIT.java:157]: The exception message may always be empty, whether or not the "alter pipe" clear it.

File: integration-test/src/test/java/org/apache/iotdb/pipe/it/IoTDBPipeAlterIT.java [hunk 2]
```diff
@@ -138,17 +185,60 @@ public void testPipeAlterFailure() {
...
+      statement.execute("alter pipe a2b modify processor ('down-sampling.interval-seconds'='2')");
+    } catch (SQLException e) {
+      fail(e.getMessage());
+    }
+
+    // check data on receiver
+    expectedResSet.clear();
+    expectedResSet.add("1000,1.0,");
+    expectedResSet.add("3000,5.0,");
+    TestUtils.assertDataOnEnv(
```
Comment (Caideyipi) [inline integration-test/src/test/java/org/apache/iotdb/pipe/it/IoTDBPipeAlterIT.java:241]: The previous progressIndex is reported, though they might be sent again when flushed to a single tsFile. However, assuring "retransmission" may not be good.

File: iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/pipe/PipeTaskInfo.java [hunk 3]
```diff
@@ -180,30 +180,36 @@ private void checkBeforeAlterPipeInternal(TAlterPipeReq alterPipeRequest) throws
...
-          String.format(
-              "Failed to alter pipe %s, nothing to alter", alterPipeRequest.getPipeName());
-      LOGGER.warn(exceptionMessage);
-      throw new PipeException(exceptionMessage);
+          copiedPipeStaticMetaFromCoordinator.getConnectorParameters().getAttribute());
+    } else if (!alterPipeRequest.isReplaceAllConnectorAttributes) {
+      alterPipeRequest.setConnectorAttributes(
+          copiedPipeStaticMetaFromCoordinator
+              .getConnectorParameters()
+              .updateEquivalentAttributes(alterPipeRequest.getConnectorAttributes())
```
Comment (Caideyipi) [inline iotdb-core/confignode/src/main/java/org/apache/iotdb/confignode/persistence/pipe/PipeTaskInfo.java:211]: What if the user wants to add some params into an existing plugin?

File: iotdb-api/pipe-api/src/main/java/org/apache/iotdb/pipe/api/customizer/parameter/PipeParameters.java [hunk 4]
```diff
@@ -307,6 +307,24 @@ public boolean isEquivalent(Object obj) {
...
+   * @param attributes Provide the key that needs to be updated along with the new value.
+   * @return this pipe parameters
+   */
+  public PipeParameters updateEquivalentAttributes(Map<String, String> attributes) {
+    attributes.forEach(
+        (k, v) ->
+            this.attributes.entrySet().stream()
+                .filter(
+                    entry ->
+                        entry.getKey().equals(k) || entry.getKey().equals(KeyReducer.reduce(k)))
```
Comment (Caideyipi) [inline iotdb-api/pipe-api/src/main/java/org/apache/iotdb/pipe/api/customizer/parameter/PipeParameters.java:323]: What if entry.getKey().equals("sink.ip") and k.equals("connector.ip")?

**Review by Caideyipi — CHANGES_REQUESTED** [review]
PTAL

File: integration-test/src/test/java/org/apache/iotdb/pipe/it/IoTDBPipeAlterIT.java (hunk 2 above)
Comment (Caideyipi) [reply integration-test/src/test/java/org/apache/iotdb/pipe/it/IoTDBPipeAlterIT.java:241]: BTW, though the progressIndex is preserved on CN, the DN inherit logic still needs to be preserved to retransmit less data... I did not notice that in the last PR...

File: iotdb-api/pipe-api/src/main/java/org/apache/iotdb/pipe/api/customizer/parameter/PipeParameters.java [hunk 5]
```diff
@@ -281,30 +283,15 @@ public String toString() {
...
-                    (oldValue, newValue) -> oldValue, TreeMap::new))
-            .toString(),
-        that.attributes.entrySet().stream()
-            .collect(
-                Collectors.toMap(
-                    entry -> KeyReducer.reduce(entry.getKey()), Entry::getValue,
-                    (oldValue, newValue) -> oldValue, TreeMap::new))
-            .toString());
+  public PipeParameters addOrReplaceAttributes(PipeParameters that) {
+    that.attributes.forEach((k, v) -> this.attributes.put(KeyReducer.reduce(k), v));
```
Comment (Caideyipi) [inline iotdb-api/pipe-api/src/main/java/org/apache/iotdb/pipe/api/customizer/parameter/PipeParameters.java:293]: This attributes may have to remove attributes with the same meaning to ensure the attributes from "that" works. Such as, "sink.ip" may have to clear “connector.ip” and "connector.host"(newly supported).

Comment (Caideyipi) [reply iotdb-api/pipe-api/src/main/java/org/apache/iotdb/pipe/api/customizer/parameter/PipeParameters.java:293]: How to tackle node-urls is tricky and may need some discussion.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/execution/config/executor/ClusterConfigTaskExecutor.java [hunk 6]
```diff
@@ -1663,10 +1663,12 @@ public SettableFuture<ConfigTaskResult> alterPipe(AlterPipeStatement alterPipeSt
 
     // Validate before alteration
     try {
-      if (!alterPipeStatement.getProcessorAttributes().isEmpty()) {
+      if (!alterPipeStatement.getProcessorAttributes().isEmpty()
+          && alterPipeStatement.isReplaceAllProcessorAttributes()) {
         PipeAgent.plugin().validateProcessor(alterPipeStatement.getProcessorAttributes());
```
Comment (Caideyipi) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/execution/config/executor/ClusterConfigTaskExecutor.java:1668]: I think it's better to try validate the "modify" as well if the pipeMeta persists in the PipeTaskMeta. If not, you can do nothing here.

File: iotdb-api/pipe-api/src/main/java/org/apache/iotdb/pipe/api/customizer/parameter/PipeParameters.java (hunk 5 above)
Comment (Caideyipi) [reply iotdb-api/pipe-api/src/main/java/org/apache/iotdb/pipe/api/customizer/parameter/PipeParameters.java:293]: A reasonable process may be like this:
1. See if reduced old attribute equals to reduced new attribute.
2. If so, remove the old one and add the new one into parameters.

**Review by SteveYurongSu — APPROVED** [review]
Greaaaaaaat

**Pull request merged**